Add EMSCRIPTEN_KEEPALIVE to N_LIB_EXPORT if emscripten

theSherwood opened the pull request:
## Summary
* `N_LIB_EXPORT` now expands to `__attribute__((used, visibility("default")))` when `__EMSCRIPTEN__` is defined.
* Compiling to wasm with emscripten won't export symbols without `__attribute__((used))`.

---
<!--- Anything before this section break (`---`) will be used as
the commit message when the PR is merged. -->

<!--
Pull Request(PR) Help

Before Merge Ensure:
* title reads like a short changelog line entry
* code includes tests and is documented
* leave the source better than before, but split out big reformats

See contributor (guide)[https://nim-works.github.io/nimskull/contributing.html]
for details, especially if you're new to this project.

Tips that make PRs easier:
* for big/impactful changes, start with chat/discussions to refine ideas
* refine the pull request message over time; don't have to nail it in one go

theSherwood committed:
Add EMSCRIPTEN_KEEPALIVE to N_LIB_EXPORT if emscripten

zerbina commented on lib/nimbase.h:
```diff
@@ -188,6 +188,8 @@ __AVR__
 
 #  ifdef __cplusplus
 #    define N_LIB_EXPORT  NIM_EXTERNC __declspec(dllexport)
+#  elif __EMSCRIPTEN__
+#    define N_LIB_EXPORT  EMSCRIPTEN_KEEPALIVE __declspec(dllexport)
```
```suggestion
```
The defines here are only meant for when targeting native Windows.

theSherwood replied:
Sounds good to me

zerbina commented on lib/nimbase.h:
```diff
@@ -220,8 +222,13 @@ __AVR__
-#  define N_LIB_EXPORT NIM_EXTERNC __attribute__((visibility("default")))
-#  define N_LIB_EXPORT_VAR  __attribute__((visibility("default")))
+#  ifdef __EMSCRIPTEN__
+#    define N_LIB_EXPORT  EMSCRIPTEN_KEEPALIVE __attribute__((visibility("default")))
+#    define N_LIB_EXPORT_VAR  EMSCRIPTEN_KEEPALIVE __attribute__((visibility("default")))
```
Since the macro comes from the `emscripten.h` header, the header would need to included here, but I'm not sure whether that's a good idea.

Personally, I'd suggest replacing `EMSCRIPTEN_KEEPALIVE` with what the macro currently expands to:

```suggestion
#    define N_LIB_EXPORT __attribute__((used, visibility("default")))
#    define N_LIB_EXPORT_VAR __attribute__((used, visibility("default")))
```

Of course, this would stop working should emscripten exported C functions start to work differently.

theSherwood replied:
Okay. Sounds good

zerbina reviewed (changes requested):
Thank you for the PR!

The define within the Windows-related section doesn't need to change, and something needs to be done about the `emscripten.h` header dependency required for the `EMSCRIPTEN_KEEPALIVE` macro (I left a suggestion).

zerbina committed:
make the CGIR goto-based and fix exception-handling bugs (#1215)

## Summary

* extend the MIR's design with goto-based control-flow primitives
* implement the new control-flow primitives for the CGIR
* make the C code generator use the new control-flow primitives
* (C backend) fix multiple bugs with `getCurrentException` returning
  the wrong exception
* (C backend) fix exceptions being leaked when aborting a `finally`
  section

### Details

The main goals of replacing the higher-level control-flow primitives in
the MIR with lower-level, goto-based ones are to:
* be able to express more complex control-flow
* make control-flow-related transformations easier and/or possible
* reduce the cost of computing the data-flow graph
* make evolving the language and MIR easier

A guiding principle was that the design should allow for enough freedom
to generate efficient code with all three code generators (which differ
significantly in control-flow-related capabilities).

To be able to implement this incrementally, as a first step, the new
control-flow primitives are only added to the CGIR, with only `cgen`
actually supporting them. So that `jsgen` and `vmgen` continue to work
without change, the legacy primitives are kept, with `cgirgen_legacy`
(a full copy of the original `cgirgen`) continuing to produce the
legacy CGIR.

### MIR

An in-depth description of syntax and behaviour is added to
`docs/mir.rst`. The gist of the design is that:
* terminators of basic-blocks become explicit (an exception being `If`)
* jump targets are directly specified on jump/fork-like operations
  (e.g., `Goto`, `CheckedCall`, etc.)
* intercepted control-flow (via `finally` sections) is also specified
  on jump/fork-like operations
* `finally` sections are *not* required to be duplicated for each
  unique control-flow passing through them
* jump targets and join points are identified by *labels*, which are
  integers IDs unique within a body

As a consequence of this design, there's no more statement-in-statement
nesting.

### CGIR & Translation

The CGIR implements all the new control-flow primitives from the MIR,
with the only difference being that the CGIR's `ContinueStmt` doesn't
track the possible exits of a `Finally` section (it's information
irrelevant to the code generators).

`cgirgen` handles translation of the old-style primitives still used by
the MIR to the new-style ones. While not strictly needed at the moment,
translation is disabled of intra-procedure unreachable code is
disabled, meaning that unreachable code is effectively dropped. This is
so that `Finally` sections that never exit normally (e.g.,
`finally: return`) can be ensured to work.

### C code generation

The CGIR is first translated into a small, low-level, C-specific IR
(all implemented in `ccgflow`). Here, the focus is on figuring out what
C code to produce for `Finally` sections; unnecessary gotos are also
eliminated.

#### Finally Handling

The previous strategy for `finally` was to:
* emit an error-state considering version of the `finally` at the end
  of the `try`
* duplicate the body of all enclosing `finally`s at `break`s and
  `return`s

Now, only a single version of the `finally` section is ever emitted. If
the `finally` section has more than one possible exit, a run-time
dispatcher is used, like so:
```c
{
  NI32 Target0_;
  L1_1_:
  Target0_ = 0;
  goto L1_;
  L1_2_:
  Target0_ = 1;
  L1_: // never jumped to from the outside
  NI32 oldNimErr0_ = *nimErr_; *nimErr_ = NIM_FALSE; // temporarily disable error mode
  // ... body of the finally ...
  *nimErr_ = oldNimErr0_;
  switch (Target0_) {
  case 0: goto L2_;
  case 1: goto L3_;
  }
}
```

Depending on the context and complexity of the code in-between, an
optimizing C compiler is able to eliminate the dispatcher and
`Target0_` assignment by directly inlining the body at a `goto L1_1_;`.

For the common case of a `Finally` section having two exits - one for
exceptional control-flow cases, and one for the -, a
`if (NIM_UNLIKELY(*nimErr_)) goto <error_label>; goto <non_error>` is
used instead of a full dispatcher. This mirrors how the exit of a
`finally` section previously looked like.

The idea behind not duplicating `Finally` sections is that it reduces
the amount of C code the compiler has to output. Nonetheless, the IR
from `ccgflow` is flexible enough to use the duplication strategy,
should this be needed again.

#### Exception Handling

Whether control-flow crosses the boundary of a (i.e., leaves it) is
encoded in the MIR/CGIR via the `Leave` item in target lists. `ccgflow`
uses this information to make sure that leaving a `Finally` via
unstructured control-flow properly aborts the active exception (if any)
and that leaving an `Except` properly pops the handler.   

#### Other

The IR from `ccgflow` then drives the rest of C code generation. Since
there's no more statement-in-statement nesting, recursion reduces
significantly.

A side-effect of the flatter representation is that less C scopes are
used (only `if`, loops, `finally`, and `except` use C scopes), which
seems to harm C compiler optimization.

Support for the legacy control-flow primitives is fully removed.

### Exception Handling Runtime

The runtime part of exception handling for the C target is partially
overhauled. For compatibility with the `csources` compiler, the new
version is guarded behind the `nimskullNewExceptionRt` condsym.

A distinction is now made between an *in-flight* exception and the
*current* exception. Previously, both were treated as being one and the
same.

An exception becomes *in-flight* once it is raised (`raiseException2`),
and stops being in-flight when it is caught (`nimCatchException`).
While in-flight, an exception can be aborted (`nimAbortException`), by
breaking out of an intercepting `finally`. On being caught by a
handler, the exception is associated with the handler and the handler
pushed to the handler stack. When control-flow leaves the handler
(`nimLeaveExcept`), the handler is popped from the stack.

The new runtime fixes the `currentException` not being updated when
breaking out of a `finally` or raising from within an `except` handler.
In addition, an exception caught by a handler can now be raised again
within the handler (e.g., `except CatchableError as e: raise e`)
without reference cycles being introduced.

zerbina committed:
refactor(mir): split `mirchangesets` into two modules (#1217)

## Summary

Split `mirchangesets` into two modules: one for the changesets
concerning only `MirTree`s and one for `MirBody` changesets. This
breaks up unnecessary module dependencies and prepares for future
extension of the `MirBody` changesets.

## Details

* the existing `Changeset` type is renamed to `TreeChangeset`
  * it only records changes for trees, and no longer has direct
    integration with `MirBuilder`
  * `insert` and `replaceMulti` provide the raw node storage
* the `MirBody`-focused API providing `MirBuilder` instances is of part
  to the new `newchangesets` module
  * inheritance (i.e., `TreeChangeset = object of RootObj`) is decided
    against because it doesn't work with `prepare(sink TreeChangeset)`
  * instead, `Changeset` stores an instance of `TreeChangeset`, with
    the `TreeChangeset` API being made available through proxy templates
* the `tmir_changesets` test is updated and renamed to
  `ttreechangesets`

`Changeset` needs to be coupled more with `MirBody` in the future.
Splitting the `MirTree`-only part into a separate type and module
makes it possible to keep using changesets where only a `MirTree` but
no body `MirBody` exists. One such case is going to be transforming/
modifying MIR constant expressions.

theSherwood committed:
Merge branch 'patch-1' of https://github.com/theSherwood/nimskull into patch-1

theSherwood committed:
Handle emscripten export symbol tags

Emscripten defines a EMSCRIPTEN_KEEPALIVE macro, which requires an emscripten.h include. So we expand N_LIB_EXPORT to include the tag that EMSCRIPTEN_KEEPALIVE uses to avoid the include.

theSherwood commented:
Closing in favor of https://github.com/nim-works/nimskull/pull/1218 because I merged from the wrong branch.
